File: CS/IssueService/IssuesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace PagedAsyncSourceSkipTokenSample {
    public static class IssuesService {
        #region helpers
        static object SyncObject = new object();
        static Lazy<IssueData[]> AllIssues = new Lazy<IssueData[]>(() => {
            var date = DateTime.Today;
            var rnd = new Random(0);
            return Enumerable.Range(0, 100000)
                .Select(i => {
                    date = date.AddSeconds(-rnd.Next(20 * 60));
                    return new IssueData(
                        subject: OutlookDataGenerator.GetSubject(),
                        user: OutlookDataGenerator.GetFrom(),
                        created: date,
                        votes: rnd.Next(100),
                        priority: OutlookDataGenerator.GetPriority());
                }).ToArray();
        });
        #endregion

        class SkipToken {
            public SkipToken(DateTime createdValue) {
                CreatedValue = createdValue;
            }
            public DateTime CreatedValue { get; private set; }
        }

        public async static Task<IssuesFetchResult> GetIssuesAsync(object skipToken, int pageSize, IssueSortOrder sortOrder) {
            await Task.Delay(300);
            IEnumerable<IssueData> issues = AllIssues.Value;

            var createdValue = skipToken != null ? ((SkipToken)skipToken).CreatedValue : default(DateTime?);
            if(sortOrder == IssueSortOrder.CreatedAscending) {
                issues = issues
                    .OrderBy(x => x.Created)
                    .Where(x => createdValue != null ? x.Created > createdValue : true);
            } else {
                issues = issues
                    .OrderByDescending(x => x.Created)
                    .Where(x => createdValue != null ? x.Created < createdValue : true);
            }

            var rows = issues.Take(pageSize).ToArray();
            var nextSkipToken = rows.Any() ? new SkipToken(rows.Last().Created) : null;
            return new IssuesFetchResult(rows, nextSkipToken);
        }
    }
}

File: VB/IssueService/IssuesService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks

Namespace PagedAsyncSourceSkipTokenSample

    Public Module IssuesService

#Region "helpers"
        Private SyncObject As Object = New Object()

        Private AllIssues As System.Lazy(Of PagedAsyncSourceSkipTokenSample.IssueData()) = New System.Lazy(Of PagedAsyncSourceSkipTokenSample.IssueData())(Function()
            Dim [date] = System.DateTime.Today
            Dim rnd = New System.Random(0)
            Return System.Linq.Enumerable.Range(0, 100000).[Select](Function(i)
                [date] = [date].AddSeconds(-rnd.[Next](20 * 60))
                Return New PagedAsyncSourceSkipTokenSample.IssueData(subject:=PagedAsyncSourceSkipTokenSample.OutlookDataGenerator.GetSubject(), user:=PagedAsyncSourceSkipTokenSample.OutlookDataGenerator.GetFrom(), created:=[date], votes:=rnd.[Next](100), priority:=PagedAsyncSourceSkipTokenSample.OutlookDataGenerator.GetPriority())
            End Function).ToArray()
        End Function)

#End Region
        Private Class SkipToken

            Private _CreatedValue As DateTime

            Public Sub New(ByVal createdValue As System.DateTime)
                Me.CreatedValue = createdValue
            End Sub

            Public Property CreatedValue As DateTime
                Get
                    Return _CreatedValue
                End Get

                Private Set(ByVal value As DateTime)
                    _CreatedValue = value
                End Set
            End Property
        End Class

        Public Async Function GetIssuesAsync(ByVal skipToken As Object, ByVal pageSize As Integer, ByVal sortOrder As PagedAsyncSourceSkipTokenSample.IssueSortOrder) As Task(Of PagedAsyncSourceSkipTokenSample.IssuesFetchResult)
            Await System.Threading.Tasks.Task.Delay(CInt((300))).ConfigureAwait(False)
            Dim issues As System.Collections.Generic.IEnumerable(Of PagedAsyncSourceSkipTokenSample.IssueData) = PagedAsyncSourceSkipTokenSample.IssuesService.AllIssues.Value
            Dim createdValue = If(skipToken IsNot Nothing, CType(skipToken, PagedAsyncSourceSkipTokenSample.IssuesService.SkipToken).CreatedValue, DirectCast(Nothing, System.DateTime?))
            If sortOrder = PagedAsyncSourceSkipTokenSample.IssueSortOrder.CreatedAscending Then
                issues = issues.OrderBy(Function(x) x.Created).Where(Function(x) If(createdValue IsNot Nothing, x.Created > createdValue, True))
            Else
                issues = issues.OrderByDescending(Function(x) x.Created).Where(Function(x) If(createdValue IsNot Nothing, x.Created < createdValue, True))
            End If

            Dim rows = issues.Take(pageSize).ToArray()
            Dim nextSkipToken = If(rows.Any(), New PagedAsyncSourceSkipTokenSample.IssuesService.SkipToken(System.Linq.Enumerable.Last(Of PagedAsyncSourceSkipTokenSample.IssueData)(rows).Created), Nothing)
            Return New PagedAsyncSourceSkipTokenSample.IssuesFetchResult(rows, nextSkipToken)
        End Function
    End Module
End Namespace
